### PR TITLE
Conventional Commits & Changelogs

### PR DESCRIPTION
**Pull request by lazd**
# Todo

- [ ] Merge #282 into this branch so we have issue templates
- [x] Create `gulp releaseBundles` task to release all bundles
- [x] Update CONTRIBUTING and release docs to refer to new conventions and processes
- [x] Make CHANGELOG template even prettier?
  - ❌ Consistently use "Changelog" (one word)
     - Blocked: https://github.com/lerna/lerna/issues/2272
  - [x] Include bold component name as scope so its listed out nicely for parseability  

# Description

This PR adds CHANGELOG.md generation based on conventional commits. In addition, we're now using yarn, bundles are no longer managed by lerna, and a new release process has been added for bundles.

### Adopt Conventional Commits

Git hooks are now included that ensure commit messages conform to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0-beta.4/) convention.

This gets us consistency and automatic releases that know exactly what version of a component to release based on the commit messages as well as automatic CHANGELOG generation.

### Generate CHANGELOGs

This PR adds CHANGELOG generation for components based on conventional commits and is able to generate CHANGELOGs for bundles as well, linking out to the individual component's CHANGELOG for details.

### Custom Spectrum CHANGELOG template

It's pretty and it knows how to link to documentation for bundles. I've asked design to weigh in to see if we can make it prettier.

### New release process for bundles

Since the version number of a bundle needs to be in step with the components inside of it, a new `gulp updateDeps` task was added to analyze the current `package.json` vs all of the components in the monorepo, update their version numbers if available, and bump the version number of the bundle appropriately. That is, if a component included in the bundle has a patch, minor, or major version bump, the bundle should get one too.

### Switch to yarn

It's faster and works based on `package.json`'s `workspaces` field. Nice.

### Codify use of `peerDependencies`

If a component relies on another component to render the example markup, it should be declared in both `devDependencies` and `peerDependencies`. 

The release will error out if:

1. A package is specified in `peerDependencies` that does not appear in `devDependencies`
2. The version of a package is specified in `devDependencies` satisfy the range defined for that package in `peerDependencies`

## Related Issues

- [SDS-2239](https://jira.corp.adobe.com/browse/SDS-2239) - Implement support for Spectrum CSS releases and changelogs with conventional commits and commitlint
- [SDS-2241](https://jira.corp.adobe.com/browse/SDS-2241) Update CONTRIBUTING.md, ISSUE_TEMPLATE.md, and PULL_REQUEST_TEMPLATE.md
- [SDS-2240](https://jira.corp.adobe.com/browse/SDS-2240) Finalize dependency declaration strategy
- [SDS-2252](https://jira.corp.adobe.com/browse/SDS-2252) Look into lerna link convert instead of lerna bootstrap

## Motivation and Context

This helps us perform releases efficiently and document what has changed.

## How Has This Been Tested?

Many fake releases.

Bundle changelog: https://git.corp.adobe.com/lawdavis/spectrum-css-lerna-test/blob/ccc/bundles/spectrum-css-compat/CHANGELOG.md

Component changelog: https://git.corp.adobe.com/lawdavis/spectrum-css-lerna-test/blob/ccc/components/calendar/CHANGELOG.md

## Screenshots (if appropriate):

Bundle changelog:
image

Component changelog:
image

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ language: node_js
 node_js:
   - "node"
 install:
-  - npm install
+  - yarn --ignore-optional
```
Comment (lazd) [inline .travis.yml]: @jianliao `yarn --ignore-optional` seems to be required to run outside of the Adobe network due to our optional dependencies... Also wondering why `@adobe/spectrum-css-workflow-icons` has optional dependencies on `@a4u` packages? I don't understand why they're there.

File: package.json [hunk 2]
```diff
@@ -3,15 +3,13 @@
...
   "scripts": {
-    "bootstrap": "npx lerna bootstrap --hoist",
     "build": "gulp -LL",
     "dev": "gulp dev",
     "clean": "gulp clean",
     "start": "npm run dev",
     "ci-all": "npm run build",
     "release": "npx lerna publish",
     "version": "gulp version",
-    "postinstall": "npm run bootstrap",
```
Comment (lazd) [inline package.json]: `yarn` does this automatically, woot!

File: package.json [hunk 3]
```diff
@@ -29,10 +27,14 @@
   },
   "license": "Apache-2.0",
   "devDependencies": {
+    "@adobe/spectrum-css-workflow-icons": "^1.0.0-beta.6",
```
Comment (lazd) [inline package.json]: This dep was moved to the top level before the `yarn install` was added to the bundle release process. It doesn't have to be here, we can move it back (that way, bundles can release with different versions of icons)

File: tools/conventional-changelog-spectrum/LICENSE.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+### ISC License
```
Comment (lazd) [inline tools/conventional-changelog-spectrum/LICENSE.md]: Not sure what to do about the fact that this was forked from https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-angular. We need to reach out to that project's owners and see if this needs to live in their monorepo...

File: tools/conventional-changelog-spectrum/templates/header.hbs [hunk 5]
```diff
@@ -0,0 +1,29 @@
+{{#if isPatch~}}
+  ##
+{{~else~}}
+  #
```
Comment (lazd) [inline tools/conventional-changelog-spectrum/templates/header.hbs]: Not sure if I like the fact that patch releases get a smaller heading... Thoughts from others?

Comment (jianliao) [reply tools/conventional-changelog-spectrum/templates/header.hbs]: I like small heading as it is more clear at the first glance.

File: lerna.json [hunk 6]
```diff
@@ -1,9 +1,18 @@
 {
   "packages": [
```
Comment (jianliao) [inline lerna.json]: You can remove `packages` as it is duplicated with `workspaces` in package.json.

File: .travis.yml (hunk 1 above)
Comment (jianliao) [reply .travis.yml]: I checked in all the workflow icon SVG files to [spectrum-css-workflow-icons repo](https://github.com/adobe/spectrum-css-workflow-icons), but the source of these icons are still coming from A4U. When A4U updates those packages, I can easily get the update by running `npm install` within adobe corp network. That is the reason I add them as optional dependencies.

For external contributor, they can't access A4U artifactory. If they want to modify the existing workflow icon, they still can fully setup the project with `npm install`. Those optional dependencies would not block them. WDYT?

File: .github/CONTRIBUTING.md [hunk 7]
```diff
@@ -47,16 +47,22 @@ git push -u origin my-awesome-new-feature
...
 ```
-npm install
+brew install yarn || curl -o- -L https://yarnpkg.com/install.sh | bash
 ```
 
 Make sure the [gulp-cli](https://github.com/gulpjs/gulp-cli) is installed globally:
 
 ```
-npm install -g gulp-cli
+yarn global all gulp-cli
```
Comment (jianliao) [inline .github/CONTRIBUTING.md]: Should be `yarn global add gulp-cli`

Comment (lazd) [reply .github/CONTRIBUTING.md]: Fixed, thanks!

**Pull request merged**